pkg/kubernetes/scripts/policy.js

```diff
@@ -32,13 +32,51 @@
         '$rootScope',
         'kubeLoader',
         'kubeMethods',
-        function($q, $rootScope, loader, methods) {
+        'kubeSelect',
+        'KubeWatch',
+        'KubeRequest',
+        'KUBE_SCHEMA',
+        function($q, $rootScope, loader, methods, select, watch, KubeRequest, KUBE_SCHEMA) {
+
+            var apiGroup;
+            var RBAC_GROUP = "rbac.authorization.k8s.io";
+            var RBAC_API = "/apis/rbac.authorization.k8s.io/v1beta1";
+            var POLICY_BINDING_API = KUBE_SCHEMA["RoleBinding"]["api"];
+            var watchPromise;
+
+            function setupRoleBinding(group) {
+                KUBE_SCHEMA["RoleBinding"]["api"] = group ? RBAC_API : POLICY_BINDING_API;
+                KUBE_SCHEMA["rolebindings"]["api"] = group ? RBAC_API : POLICY_BINDING_API;
+                apiGroup = group;
+                expireSAR(null);
+                expireWhoCan(null);
+                return group ? "rolebindings" : "policybindings";
+            }
+
+            function ensureWatchType() {
+                if (!watchPromise) {
+                    watchPromise = new KubeRequest("GET", "/oapi/v1")
+                        .then(function(response) {
+                            var data = response.data || {};
+                            var i, l = data.resources || [];
+                            for(i = 0; i < l.length; i++ ) {
+                                if (l[i].kind == "PolicyBinding")
+                                    return setupRoleBinding();
+                            }
+                            return setupRoleBinding(RBAC_GROUP);
```

Comment:
So this assumes that if there is no PolicyBinding resource (as in < 3.7) there will be the RBAC API. Does that appear in the `/oapi/v1` resource list? If so, would it make sense to explicitly test for it in the iteration, or do you consider it as the current "last/opportunistic fallback"? (I. e. is that beta RBAC API supposed to become stable and long-term supported with 3.7).

Reply:
The /oapi/v1 has an rolebinding endpoint. But it is not watchable. Additionally those api objects are different from the rbac ones, different selfLinks and versions and all that. However changes made to the rbac objects are reflected there and visa versa.

However in 3.6 where rbac and policybindings are both available, changes made to the rbac objects do not result in changes to the policybindings.

So when there are no policy bindings it is safe to use rbac. Otherwise we stick with policy bindings to make sure we are not missing things that are only set in the policybinding.